### PR TITLE
Fix #315 #342

### LINKED ISSUE #315: Use Arbitrary index in AtTests / IndexTests
At the moment, `AtTests` /  `IndexTests` use a constant index, it would be much more useful if the index was random. See https://github.com/julien-truffaut/Monocle/blob/master/law/src/main/scala/monocle/law/discipline/function/AtTests.scala

### PR DESCRIPTION
**Pull request by pvillega**
Fix issue in Bits where singleBit for Long was shitfing bits incorrectly

Add random index for IndexTests/OptionalTests using ScalaCheck

### CODE REVIEW

File: build.sbt [hunk 1]
```diff
@@ -43,6 +43,7 @@ lazy val buildSettings = Seq(
 lazy val scalaz     = Def.setting("org.scalaz"      %%% "scalaz-core" % "7.2.1")
 lazy val shapeless  = Def.setting("com.chuusai"     %%% "shapeless"   % "2.3.0")
 lazy val refinedDep = Def.setting("eu.timepit"      %%% "refined"     % "0.3.7")
+lazy val refinedScalaCheckDep = Def.setting("eu.timepit"      %%% "refined-scalacheck"  % "0.3.7")
```
Comment (julien-truffaut) [inline build.sbt]: can you add "test" modifier

**Comment by pvillega** [comment]
Updated! I had to move the dependency for it to compile, let me know if there's anything else you want changed :)

File: law/shared/src/main/scala/monocle/law/discipline/function/AtTests.scala [hunk 2]
```diff
@@ -7,18 +7,17 @@ import org.scalacheck.Arbitrary
...
-                                    sEq: Equal[S], sArb: Arbitrary[S],
-                                   evAt: At[S, I, A]): RuleSet =
-    new SimpleRuleSet("At", LensTests(at(i)).props: _*)
+  def apply[S, I, A](implicit aEq: Equal[A], aArb: Arbitrary[A],
+                     sEq: Equal[S], sArb: Arbitrary[S],
+                     iArb: Arbitrary[I], evAt: At[S, I, A]): RuleSet = {
+    new SimpleRuleSet("At", LensTests(at(_: I)).props: _*)
+  }
 
   def defaultIntIndex[S, A](implicit aEq: Equal[A], aArb: Arbitrary[A],
```
Comment (julien-truffaut) [inline law/shared/src/main/scala/monocle/law/discipline/function/AtTests.scala]: we can probably remove `defaultIntIndex`

File: law/shared/src/main/scala/monocle/law/discipline/function/AtTests.scala [hunk 3]
```diff
@@ -7,18 +7,17 @@ import org.scalacheck.Arbitrary
...
-import scalaz.std.option._
 
 object AtTests extends Laws {
 
-  def apply[S, I, A](i: I)(implicit aEq: Equal[A], aArb: Arbitrary[A],
-                                    sEq: Equal[S], sArb: Arbitrary[S],
-                                   evAt: At[S, I, A]): RuleSet =
-    new SimpleRuleSet("At", LensTests(at(i)).props: _*)
+  def apply[S, I, A](implicit aEq: Equal[A], aArb: Arbitrary[A],
+                     sEq: Equal[S], sArb: Arbitrary[S],
```
Comment (julien-truffaut) [inline law/shared/src/main/scala/monocle/law/discipline/function/AtTests.scala]: do you mind to maintain the vertical alignment

File: law/shared/src/main/scala/monocle/law/discipline/function/IndexTests.scala [hunk 4]
```diff
@@ -10,14 +10,14 @@ import scalaz.Equal
...
-  def apply[S, I, A](i: I)(implicit aEq: Equal[A], aArb: Arbitrary[A],
-                                    sEq: Equal[S], sArb: Arbitrary[S],
-                                evIndex: Index[S, I, A]): RuleSet =
-    new SimpleRuleSet("Index", OptionalTests(index(i)).props: _*)
+  def apply[S, I, A](implicit aEq: Equal[A], aArb: Arbitrary[A],
+                     sEq: Equal[S], sArb: Arbitrary[S],
+                     iArb: Arbitrary[I], evIndex: Index[S, I, A]): RuleSet =
+    new SimpleRuleSet("Index", OptionalTests(index(_ : I)).props: _*)
 
   def defaultIntIndex[S, A](implicit aEq: Equal[A], aArb: Arbitrary[A],
```
Comment (julien-truffaut) [inline law/shared/src/main/scala/monocle/law/discipline/function/IndexTests.scala]: same

**Comment by julien-truffaut** [comment]
Thank you very much, it is awesome to write tests that actually find bugs :)

**Comment by pvillega** [comment]
No, thank you! It was fun and I learned :) 

I've removed the methods, updated code to user the 'proper' apply way, and fixed the styling (hopefully). Let me know if you want me to tweak something else, as IntelliJ may do stuff I don't realise... :eh:

**Comment by julien-truffaut** [comment]
:+1:

**Pull request merged**